File: 0002-add-two-numbers/0002-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2, int carry=0) {
        if(l1==nullptr && l2==nullptr){
            if(carry!=0)    return new ListNode(carry);
            return l1; 
        }

        int sum = carry;
        ListNode* next_l1;
        ListNode* next_l2;
        if(l1!=nullptr){
            sum+=l1->val; 
            next_l1 = l1->next;
        }
        else {l1 = l2;}
        
        if(l2!=nullptr){
            sum+=l2->val; 
            next_l2 = l2->next;
        }
        l1->val = sum % 10; 
        l1->next = addTwoNumbers(next_l1, next_l2, sum/10);
        return l1;
        
    }
};

File: 0002-add-two-numbers/0002-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        l1_ptr, l2_ptr = l1, l2
        carry = 0
        sum_head = ListNode(0)
        sum_ptr = sum_head
        while l1_ptr or l2_ptr or carry:
            sum_ = carry
            if l1_ptr:
                sum_ += l1_ptr.val
                l1_ptr = l1_ptr.next
            if l2_ptr:
                sum_ += l2_ptr.val
                l2_ptr = l2_ptr.next
                
            digit = ListNode(sum_ % 10)
            sum_ptr.next = digit
            sum_ptr = sum_ptr.next
            carry = sum_ // 10
        return sum_head.next
            
            
            
        